### PR TITLE
Collateralization Limits Part 2

### PR DESCRIPTION
**Pull request by 0xripleys**
Summary:
1. allow lending market owner or risk authority to mark obligations as closeable. 
2. Modify liquidate instruction to handle liquidating "closeable" obligations. These obligations are closed without penalty.

### CODE REVIEW

File: token-lending/program/src/processor.rs [hunk 1]
```diff
@@ -3115,6 +3136,84 @@ pub fn process_resize_reserve(_program_id: &Pubkey, accounts: &[AccountInfo]) ->
     Ok(())
 }
 
+/// process mark obligation as closable
+pub fn process_mark_obligation_as_closeable(
```
Comment (nope-finance) [inline token-lending/program/src/processor.rs]: ig remove expiry time and change this to "set obligation close ability status"

File: token-lending/program/src/processor.rs [hunk 2]
```diff
@@ -1081,7 +1086,13 @@ fn process_refresh_obligation(program_id: &Pubkey, accounts: &[AccountInfo]) ->
 
     obligation.last_update.update_slot(clock.slot);
 
-    update_borrow_attribution_values(&mut obligation, &accounts[1..], false)?;
+    let any_borrow_attribution_limit_exceeded =
+        update_borrow_attribution_values(&mut obligation, &accounts[1..], false)?;
+
+    // unmark obligation as closable after it's been liquidated enough times
+    if obligation.is_closeable(clock.slot) && !any_borrow_attribution_limit_exceeded {
```
Comment (nope-finance) [inline token-lending/program/src/processor.rs]: close vs open attribution limits

File: token-lending/program/src/processor.rs [hunk 3]
```diff
@@ -3115,6 +3136,84 @@ pub fn process_resize_reserve(_program_id: &Pubkey, accounts: &[AccountInfo]) ->
...
+    if &obligation.lending_market != lending_market_info.key {
+        msg!("Obligation lending market does not match the lending market provided");
+        return Err(LendingError::InvalidAccountInput.into());
+    }
+    if obligation.last_update.is_stale(clock.slot)? {
+        msg!("Obligation is stale and must be refreshed");
+        return Err(LendingError::ObligationStale.into());
+    }
+
+    if &lending_market.risk_authority != risk_authority_info.key {
```
Comment (nope-finance) [inline token-lending/program/src/processor.rs]: or owner

File: token-lending/sdk/src/state/reserve.rs [hunk 4]
```diff
@@ -338,8 +338,16 @@ impl Reserve {
...
     /// the value will be in range [0, MAX_BONUS_PCT]
-    pub fn calculate_bonus(&self, obligation: &Obligation) -> Result<Decimal, ProgramError> {
+    pub fn calculate_bonus(
+        &self,
+        obligation: &Obligation,
+        current_slot: Slot,
+    ) -> Result<Decimal, ProgramError> {
         if obligation.borrowed_value < obligation.unhealthy_borrow_value {
+            if obligation.is_closeable(current_slot) {
+                return Ok(Decimal::zero());
```
Comment (nope-finance) [inline token-lending/sdk/src/state/reserve.rs]: i think gotta remove the protocol liquidation fee too

**Review by nope-finance — CHANGES_REQUESTED** [review]

**Comment by 0xripleys** [comment]
actually i don't think we need the init flag. the obligation collateral's borrow attribution value on upgrade will be zero, so the subtraction is useless anyways.

File: token-lending/program/src/processor.rs [hunk 5]
```diff
@@ -3185,13 +3192,14 @@ pub fn process_mark_obligation_as_closeable(
         return Err(LendingError::ObligationStale.into());
     }
 
-    if &lending_market.risk_authority != risk_authority_info.key {
+    if &lending_market.risk_authority != signer_info.key && &lending_market.owner != signer_info.key
+    {
         msg!("Lending market risk authority does not match the risk authority provided");
```
Comment (nope-finance) [inline token-lending/program/src/processor.rs]: update msg

File: token-lending/sdk/src/state/reserve.rs [hunk 6]
```diff
@@ -338,13 +338,9 @@ impl Reserve {
 
     /// Calculate bonus as a percentage
     /// the value will be in range [0, MAX_BONUS_PCT]
-    pub fn calculate_bonus(
-        &self,
-        obligation: &Obligation,
-        current_slot: Slot,
-    ) -> Result<Decimal, ProgramError> {
```
Comment (nope-finance) [inline token-lending/sdk/src/state/reserve.rs]: lets just return the protocol liquidation fee here and save self from re-serializing later

**Pull request merged**